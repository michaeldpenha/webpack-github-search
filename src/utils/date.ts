import formatDistanceToNow from 'date-fns/formatDistanceToNow';

/**
 * This function will return formatted date string
 * @param createdAt {string}
 * @returns {string} string with formatted date
 */
export const formatDistance = (createdAt: string): string =>
  !isNaN(Date.parse(createdAt))
    ? formatDistanceToNow(new Date(createdAt), { addSuffix: true })
    : '';
