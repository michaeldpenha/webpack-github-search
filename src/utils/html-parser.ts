import parse from 'html-react-parser';

/**
 * This function will parse Html string and will return HTML elements using html-react-parser
 * @param val
 * @returns {element | string}
 */
export const parseHTML = (val: string): string | JSX.Element | JSX.Element[] => parse(val);
