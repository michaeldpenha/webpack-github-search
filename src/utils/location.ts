import { History, Location } from 'history';

/**
 * This function will update the url params without refresh
 * @param location location object
 * @param value which needs to be updated
 * @param history history object
 * @param key param key
 * @param append whether to append or to delete
 * @return {void}
 */

export const updaeUrlQueryParams = (
  location: Location,
  value: string,
  history: History,
  key: string,
  append?: boolean,
) => {
  const params: URLSearchParams = new URLSearchParams(location.search);
  params.delete(key);
  if (append) {
    params.append(key, value);
  }
  history.replace({
    search: params.toString(),
  });
};
