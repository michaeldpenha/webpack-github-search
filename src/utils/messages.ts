import { MESSAGES } from 'src/messages';

export const formatMessageText = (key: string, values?: string[]): string => {
  const localeText: string = MESSAGES[key];
  if (!values?.length) {
    return localeText;
  }

  return values.reduce(
    (result: string, value: string, index: number) => result.replace(`{${index}}`, value),
    localeText,
  );
};
