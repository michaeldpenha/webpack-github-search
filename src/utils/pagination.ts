import { IssueComment, SearchResultItemEdge, PageInfo } from 'types/schema';

interface IData {
  edges: Array<IssueComment | SearchResultItemEdge>;
  pageInfo: PageInfo;
  totalCount?: number;
}
/**
 * This function is basically merging edges object for pagination
 * @param currentData
 * @param newData
 * @returns {IData}Merged edges data along with pageInfo
 */
export const mergeData = (currentData: IData, newData: IData): IData => {
  const result: IData = {
    pageInfo: newData.pageInfo,
    edges: currentData ? [...currentData.edges, ...newData.edges] : newData.edges,
    totalCount: newData.totalCount,
  };

  return result;
};
