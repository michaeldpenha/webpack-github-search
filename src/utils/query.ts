import { DEFAULT_ISSUE_SEARCH_QUERY } from 'src/constants';

/**
 * Funciton will return query string which will be used in search query
 * @param params URL search query object
 * @returns {string} query string
 */
export const prepareSearchQuery = (params: URLSearchParams): string => {
  const queryParams = {
    type: DEFAULT_ISSUE_SEARCH_QUERY.type,
    language: params.get('langauge') || DEFAULT_ISSUE_SEARCH_QUERY.language,
    state: params.get('state') || DEFAULT_ISSUE_SEARCH_QUERY.state,
  };

  if (params.get('search')) {
    queryParams['in'] = `title,body ${params.get('search')}`;
  }

  if (params.get('repo')) {
    queryParams['repo'] = params.get('repo');
  }

  const result = Object.keys(queryParams).reduce((prev: string, key: string) => {
    return `${prev} ${key}:${queryParams[key]}`;
  }, '');
  return result;
};
