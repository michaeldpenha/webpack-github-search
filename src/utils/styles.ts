import { BASE_FONT_SIZE } from '../constants';

/**
 * Funciton will return the size | padding | margin of the elements
 * @param size number
 * @returns {string} with rem
 */
export const getStylesSize = (size: number): string => {
  return `${size / BASE_FONT_SIZE}rem`;
};
