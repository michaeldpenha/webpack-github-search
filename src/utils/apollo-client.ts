import {
  ApolloClient,
  ApolloLink,
  HttpLink,
  InMemoryCache,
  NormalizedCacheObject,
} from '@apollo/client';

import { config } from '../../config';
import { mergeData } from './pagination';

const { githubGraphQlEndPoint: uri } = config;

const authMiddleware = (): ApolloLink =>
  new ApolloLink((operation, forward) => {
    operation.setContext({
      headers: {
        authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
      },
    });
    return forward(operation);
  });

const cache: InMemoryCache = new InMemoryCache({
  possibleTypes: {
    SearchResultItem: ['Issue', 'Repository'],
  },
  typePolicies: {
    Query: {
      fields: {
        search: {
          keyArgs: ['query'],
          merge(existing, incoming) {
            return mergeData(existing, incoming);
          },
          read(existing) {
            return existing;
          },
        },
      },
    },
    Issue: {
      keyFields: ['id', 'number'],
      fields: {
        comments: {
          keyArgs: [],
          merge(existing, incoming) {
            return mergeData(existing, incoming);
          },
          read(existing) {
            return existing;
          },
        },
      },
    },
    Repository: {
      keyFields: ['name', 'id'],
    },
    IssueComment: {
      keyFields: ['id'],
    },
    User: {
      keyFields: ['login'],
    },
  },
});

const httpLink: HttpLink = new HttpLink({
  uri,
});

// const persistedHttpLink: ApolloLink = createPersistedQueryLink({
//   useGETForHashedQueries: true,
//   sha256,
// }).concat(httpLink);

export const apolloClient: ApolloClient<NormalizedCacheObject> = new ApolloClient({
  cache,
  link: authMiddleware().concat(httpLink),
});
