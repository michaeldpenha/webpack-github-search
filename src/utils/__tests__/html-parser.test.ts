import { parseHTML } from 'utils/html-parser';

describe('html parser', () => {
  it('should return string', () => {
    expect(parseHTML('test')).toBe('test');
  });

  it('should return div element', () => {
    expect(parseHTML('<div>test</div>')).toMatchSnapshot();
  });
});
