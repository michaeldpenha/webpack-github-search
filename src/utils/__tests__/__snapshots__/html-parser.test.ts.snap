// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html parser should return div element 1`] = `
<div>
  test
</div>
`;
